### PR TITLE
Extract SLG solver into its own crate

### PR DESCRIPTION
**Pull request by nikomatsakis**
The history is long and tortured, but it worked out pretty well!

The `chalk_src/context/mod.rs` crate defines the interface to the rest of the world. Can probably be simplified somewhat further. Should be reusable in rustc though.

cc @scalexm

### CODE REVIEW

File: chalk-slg/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+impl<C: Context> PartialEq for Literal<C> {
+    fn eq(&self, other: &Literal<C>) -> bool {
+        match (self, other) {
+            (Literal::Positive(goal1), Literal::Positive(goal2))
+            | (Literal::Negative(goal1), Literal::Negative(goal2)) => goal1 == goal2,
+
+            _ => false,
+        }
+    }
+}
```
Comment (scalexm) [inline chalk-slg/src/lib.rs]: Is there a reason for not `#[derive]`-ing these impls?

Comment (nikomatsakis) [reply chalk-slg/src/lib.rs]: Gah. I'm torn. On the one hand, they are not *needed*, and it requires deriving a bunch of ord and crap elsewhere that are also not *needed*.

On the *other* hand, writing the impls by hand is a real pain. After having tried that for long enough, I'm maybe inclined to just add the derives.

Comment (nikomatsakis) [reply chalk-slg/src/lib.rs]: actually I was being silly. Fixed.

File: src/fold/mod.rs [hunk 2]
```diff
@@ -365,7 +365,7 @@ macro_rules! copy_fold {
         impl ::fold::Fold for $t {
             type Result = Self;
             fn fold_with(&self,
-                         _folder: &mut ::fold::Folder,
+                         _folder: &mut dyn (::fold::Folder),
```
Comment (nikomatsakis) [inline src/fold/mod.rs]: @Manishearth -- well, I hit the edge case now =)

Comment (Manishearth) [reply src/fold/mod.rs]: Oh, that edge case is fine, it's the other way around. Will people ever be using `dyn::foo` as a path?

Comment (nikomatsakis) [reply src/fold/mod.rs]: > Will people ever be using dyn::foo as a path?

Who knows...

**Pull request merged**